### PR TITLE
[MI-1987]: Integrated project list UI

### PR DESCRIPTION
**Pull request by avas27JTG**
Integrated project list UI with:
- Loader
- Cards to show projects
- Unlink confirmation dialog

Synced branches

image

image

image

1st pass review is already done here https://github.com/Brightscout/mattermost-plugin-azure-devops/pull/14

### CODE REVIEW

File: webapp/src/components/modal/index.tsx [hunk 1]
```diff
@@ -9,22 +9,23 @@ import ModalSubTitleAndError from './subComponents/modalSubtitleAndError';
 
 type ModalProps = {
     show: boolean;
-    onHide: () => void;
+    onHide?: () => void;
```
Comment (sooraj-shukla) [inline webapp/src/components/modal/index.tsx:12]: This should be a required prop.

File: webapp/src/components/modal/index.tsx [hunk 2]
```diff
@@ -54,8 +55,7 @@ const Modal = ({show, onHide, showCloseIconInHeader = true, children, title, sub
                 onConfirm={onConfirm}
                 cancelBtnText={cancelBtnText}
                 confirmBtnText={confirmBtnText}
-                confirmDisabled={confirmDisabled}
```
Comment (sooraj-shukla) [inline webapp/src/components/modal/index.tsx:57]: Why is this removed?

File: webapp/src/reducers/projectDetails/index.ts [hunk 3]
```diff
@@ -1,24 +1,24 @@
...
+            state.projectName = action.payload.projectName;
+            state.organizationName = action.payload.organizationName;
         },
         resetProjectDetails: (state: ProjectDetails) => {
-            state.id = '';
-            state.title = '';
-            state.organization = '';
+            state.projectID = '';
+            state.projectName = '';
+            state.organizationName = '';
```
Comment (sooraj-shukla) [inline webapp/src/reducers/projectDetails/index.ts:21]: Can't we do `state = initialState` here?

File: webapp/src/services/index.ts [hunk 4]
```diff
@@ -11,22 +11,29 @@ const pluginApi = createApi({
     baseQuery: fetchBaseQuery({baseUrl: Utils.getBaseUrls().pluginApiBaseUrl}),
     tagTypes: ['Posts'],
     endpoints: (builder) => ({
-        [Constants.pluginApiServiceConfigs.createTask.apiServiceName]: builder.query<void, CreateTaskPayload>({
+        [Constants.pluginApiServiceConfigs.createTask.apiServiceName]: builder.query<void, APIRequestPayload>({
```
Comment (sooraj-shukla) [inline webapp/src/services/index.ts:14]: Since, this API is for `createTask`, shouldn't payload type be `CreateTaskPayload` rather than `APIRequestPayload`.

File: webapp/src/services/index.ts [hunk 5]
```diff
@@ -11,22 +11,29 @@ const pluginApi = createApi({
...
             query: (payload) => ({
                 headers: {[Constants.HeaderMattermostUserID]: Cookies.get(Constants.MMUSERID)},
                 url: Constants.pluginApiServiceConfigs.createTask.path,
                 method: Constants.pluginApiServiceConfigs.createTask.method,
                 body: payload,
             }),
         }),
-        [Constants.pluginApiServiceConfigs.createLink.apiServiceName]: builder.query<void, LinkPayload>({
+        [Constants.pluginApiServiceConfigs.createLink.apiServiceName]: builder.query<void, APIRequestPayload>({
```
Comment (sooraj-shukla) [inline webapp/src/services/index.ts:22]: Same as above

**Review by sooraj-shukla — CHANGES_REQUESTED** [review]

File: webapp/src/services/index.ts (hunk 4 above)
Comment (avas27JTG) [reply webapp/src/services/index.ts:14]: we have created one type `APIRequestPayload` which is union of different types, if you see type file we have `type APIRequestPayload = CreateTaskPayload | LinkPayload | void;` this is to generalise changes

File: webapp/src/reducers/projectDetails/index.ts (hunk 3 above)
Comment (avas27JTG) [reply webapp/src/reducers/projectDetails/index.ts:21]: no, and we should not do this even in the normal redux flow because object is passed by reference so after manipulation `initialState` does not have initial values i.e. empty data anymore

File: webapp/src/components/modal/index.tsx (hunk 2 above)
Comment (avas27JTG) [reply webapp/src/components/modal/index.tsx:57]: this has been fixed here https://github.com/Brightscout/mattermost-plugin-azure-devops/pull/40/files

**Review by sooraj-shukla — APPROVED** [review]

File: webapp/src/components/emptyState/index.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    subTitle?: {
+        text: string
+        slashCommand?: string
+    },
+    buttonText?: string,
+    buttonAction?: (event: React.SyntheticEvent) => void;
+}
+
+const EmptyState = ({title, subTitle, buttonText, buttonAction}: EmptyStatePropTypes) => {
+    return (
```
Comment (manojmalik20) [inline webapp/src/components/emptyState/index.tsx:16]:
```suggestion
const EmptyState = ({title, subTitle, buttonText, buttonAction}: EmptyStatePropTypes) => (
```

File: webapp/src/components/modal/confirmationModal/index.tsx [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    isOpen: boolean
+    title: string
+    description: string
+    confirmBtnText: string
+    onHide: () => void
+    onConfirm?: () => void
+}
+
+const ConfirmationModal = ({isOpen, title, confirmBtnText, description, onHide, onConfirm}: ConfirmationModalProps) => {
+    return (
```
Comment (manojmalik20) [inline webapp/src/components/modal/confirmationModal/index.tsx:15]:
```suggestion
const ConfirmationModal = ({isOpen, title, confirmBtnText, description, onHide, onConfirm}: ConfirmationModalProps) => (
```

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 8]
```diff
@@ -1,47 +1,76 @@
...
 
+    const handleOpenLinkProjectModal = () => {
+        dispatch(showLinkModal([]));
+    };
+
+    useEffect(() => {
+        usePlugin.makeApiRequest(plugin_constants.pluginApiServiceConfigs.getAllLinkedProjectsList.apiServiceName);
+    }, []);
+
+    const data = usePlugin.getApiState(plugin_constants.pluginApiServiceConfigs.getAllLinkedProjectsList.apiServiceName).data;
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectList/index.tsx:31]:
```suggestion
    const {data} = usePlugin.getApiState(plugin_constants.pluginApiServiceConfigs.getAllLinkedProjectsList.apiServiceName);
```

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 9]
```diff
@@ -1,47 +1,76 @@
...
+            {
+                usePlugin.getApiState(plugin_constants.pluginApiServiceConfigs.getAllLinkedProjectsList.apiServiceName).isLoading && (
+                    <LinearLoader/>
+                )
+            }
+            {
+                usePlugin.getApiState(plugin_constants.pluginApiServiceConfigs.getAllLinkedProjectsList.apiServiceName).isSuccess &&
+                data &&
+                (
+                    data?.length > 0 ?
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectList/index.tsx:55]:
```suggestion
                    data.length ?
```

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 10]
```diff
@@ -1,47 +1,76 @@
...
+                        data?.map((item) => (
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectList/index.tsx:56]:
```suggestion
                        data.map((item) => (
```
You've already added a check for `data` above.

File: webapp/src/selectors/index.tsx [hunk 11]
```diff
@@ -5,3 +5,7 @@ const pluginPrefix = `plugins-${plugin_constants.pluginId}`;
 export const getprojectDetailsState = (state: any) => {
     return state[pluginPrefix].projectDetailsSlice;
 };
+
+export const getRhsState = (state: any): {isSidebarOpen: boolean} => {
```
Comment (manojmalik20) [inline webapp/src/selectors/index.tsx:9]: Why is the type of `state` `any` here?

**Review by manojmalik20 — COMMENTED** [review]
Try to fix the CI if possible.

File: webapp/src/selectors/index.tsx (hunk 11 above)
Comment (avas27JTG) [reply webapp/src/selectors/index.tsx]: this was left for the enhancements part but as we are now doing that, fixed it now

File: webapp/src/selectors/index.tsx [hunk 12]
```diff
@@ -1,11 +1,7 @@
...
-const pluginPrefix = `plugins-${plugin_constants.pluginId}`;
-
-export const getprojectDetailsState = (state: any) => {
-    return state[pluginPrefix].projectDetailsSlice;
+export const getprojectDetailsState = (state: ReduxState) => {
+    return state['plugins-mattermost-plugin-azure-devops'].projectDetailsSlice;
 };
 
-export const getRhsState = (state: any): {isSidebarOpen: boolean} => {
+export const getRhsState = (state: ReduxState): {isSidebarOpen: boolean} => {
```
Comment (manojmalik20) [inline webapp/src/selectors/index.tsx:5]:
```suggestion
export const getRhsState = (state: GlobalState): {isSidebarOpen: boolean} => {
```

File: webapp/src/types/common/index.d.ts [hunk 13]
```diff
@@ -12,7 +12,12 @@ type PluginApiService = {
     apiServiceName: ApiServiceName
 }
 
-type PluginState = {
+interface ReduxState extends GlobalState {
+    views: {
+        rhs: {
+            isSidebarOpen: boolean
+        }
+    }
```
Comment (manojmalik20) [inline webapp/src/types/common/index.d.ts:20]: There is no need for this. You can directly use the GlobalState to access this `views` state.

**Review by manojmalik20 — APPROVED** [review]

File: webapp/src/types/common/index.d.ts (hunk 13 above)
Comment (avas27JTG) [reply webapp/src/types/common/index.d.ts:20]: actually, there is no definition for this in MM redux type

```
export declare type GlobalState = {
    entities: {
        general: GeneralState;
        users: UsersState;
        teams: TeamsState;
        channels: ChannelsState;
        posts: PostsState;
        bots: {
            accounts: Dictionary<Bot>;
        };
        preferences: {
            myPreferences: {
                [x: string]: PreferenceType;
            };
        };
        admin: AdminState;
        jobs: JobsState;
        search: SearchState;
        integrations: IntegrationsState;
        files: FilesState;
        emojis: EmojisState;
        typing: Typing;
        roles: {
            roles: {
                [x: string]: Role;
            };
            pending: Set<string>;
        };
        schemes: SchemesState;
        gifs: any;
        groups: GroupsState;
        channelCategories: ChannelCategoriesState;
    };
    errors: Array<any>;
    requests: {
        channels: ChannelsRequestsStatuses;
        general: GeneralRequestsStatuses;
        posts: PostsRequestsStatuses;
        teams: TeamsRequestsStatuses;
        users: UsersRequestsStatuses;
        admin: AdminRequestsStatuses;
        files: FilesRequestsStatuses;
        roles: RolesRequestsStatuses;
        jobs: JobsRequestsStatuses;
    };
    websocket: {
        connected: boolean;
        lastConnectAt: number;
        lastDisconnectAt: number;
    };
};

```

Comment (manojmalik20) [reply webapp/src/types/common/index.d.ts:20]: Then you should update the version of `mattermost-redux` in `package.json`. @avas27JTG
